src/token.rs

```diff
@@ -441,215 +306,574 @@ impl Token {
         return Ok(());
     }
 
-    pub fn initialize(&mut self, pin: &Vec<u8>, label: &Vec<u8>) -> CK_RV {
-        let ret = if self.is_initialized() {
-            self.login(CKU_SO, pin)
-        } else {
-            self.so_login.set_pin(&self.info, pin)
+    fn fetch_pin_object(&mut self, uid: &str) -> KResult<Object> {
+        let obj = match self.storage.fetch_by_uid(&uid.to_string()) {
+            Ok(o) => o,
+            Err(e) => match e {
+                KError::NotFound(_) => {
+                    return err_rv!(CKR_USER_PIN_NOT_INITIALIZED);
+                }
+                KError::RvError(e) => return err_rv!(e.rv),
+                _ => return err_rv!(CKR_GENERAL_ERROR),
+            },
         };
-        if ret != CKR_OK {
-            return ret;
+        if obj.get_attr_as_ulong(CKA_CLASS)? != CKO_SECRET_KEY {
+            return err_rv!(CKR_GENERAL_ERROR);
         }
-        self.so_login.logged_in = false;
-        match self.storage.reinit() {
-            Ok(()) => {
-                self.handles = Handles::new();
-            }
-            Err(_) => return CKR_GENERAL_ERROR,
+        if obj.get_attr_as_ulong(CKA_KEY_TYPE)? != CKK_GENERIC_SECRET {
+            return err_rv!(CKR_GENERAL_ERROR);
         }
-        self.session_objects.clear();
+        Ok(obj)
+    }
 
-        /* add pin to so_object */
-        if self
-            .store_pin_object(
-                SO_PIN_UID.to_string(),
-                "SO PIN".to_string(),
-                pin.clone(),
-            )
-            .is_err()
-        {
-            return CKR_GENERAL_ERROR;
-        }
+    fn store_pin_object(
+        &mut self,
+        uid: String,
+        label: String,
+        wrapped: Vec<u8>,
+    ) -> KResult<()> {
+        match self.storage.fetch_by_uid(&uid) {
+            Ok(o) => {
+                let mut obj = o.clone();
+                obj.set_attr(attribute::from_string(CKA_LABEL, label))?;
+                obj.set_attr(attribute::from_bytes(CKA_VALUE, wrapped))?;
+                obj.set_attr(attribute::from_ulong(KRA_LOGIN_ATTEMPTS, 0))?;
+                self.storage.store(&uid, obj)?;
```

Comment:
So this will be completely implemented outside of the storage layer? This means the `enc` column can be removed from the database schema, as well as from todos in the sqlite backend?

https://github.com/latchset/kryoptic/blob/eee73043e7232714f17c857e034e1bb65de8acaa/src/storage/sqlite.rs#L43

Reply:
That's a good q. On the one hand I think it would be nice to mark encrypted vs clear text values somehow. OTOH the solution i have now is cleaner than trying to cram it at the db layer.

Should I just drop that column for now?

Reply:
If it is not used, I would drop it.
It might become needed when we will need some backward compatibility, we will change what is encrypted or how. But for now, it sits there unused and I would argue it might be confusing for readers.

Reply:
dropping